omniduct/databases/hiveserver2.py

```diff
@@ -135,11 +179,101 @@ def _log_status(self, cursor, log_offset=0):
+    def _push(self, df, table, if_exists='fail', schema=None, partition_cols=None,
+              partition_clause=None, sep=chr(1), table_props=None, dtype_overrides=None, **kwargs):
+        """
+        If `.push_using_hive_cli` is `True`, a `CREATE TABLE` statement will be
+        automatically generated based on the datatypes of the DataFrame (unless
+        overwritten by `dtype_overrides`). The `DataFrame` will then be exported to
+        a CSV compatible with Hive and uploaded (if necessary) to the remote,
+        before being loaded into Hive using a `LOAD DATA LOCAL INFILE ...`
+        query using the `hive` cli executable.
+
+        If `.push_using_hive_cli` is `False`, an attempt will be made to push
+        the `DataFrame` to Hive using `pandas.DataFrame.to_sql` and the SQLAlchemy
+        binding provided by `pyhive` and `impyla`. This is typically much slower,
+        and may not be supported on older versions of Hive.
+
+        Additional Parameters:
+            schema (str): The schema into which the table should be pushed. If
+                not specified, the schema will be set to your username.
+            **kwargs (dict): Additional arguments to send to `pandas.DataFrame.to_sql`.
+
+        Further Parameters for CLI method (ignored for pandas method):
+            partition_cols (list<str>): Names of columns that are or should be used
+                to partition the table.
+            partition_clause (dict): A mapping of column names to values that specify
+                the partition into which the provided data should be uploaded.
+            sep (str): Field delimiter for data (defaults to CTRL-A, or \\u0001).
+            table_props (dict): Properties to set on any newly created tables
+                (extends `.default_table_props`).
+            dtype_overrides (dict): Mapping of column names to Hive datatypes to
+                use instead of default mapping.
+        """
+        schema = schema or self.username
+
+        # Try using SQLALchemy method
+        if not self.push_using_hive_cli:
+            try:
+                return df.to_sql(name=table, con=self._sqlalchemy_engine, index=False, if_exists=if_exists, schema=schema, **kwargs)
+            except Exception as e:
+                raise RuntimeError("Push unsuccessful. Your version of Hive may be too old to support the `INSERT` keyword. Original exception was: {}".format(e.args[0]))
+
+        # Try using Hive CLI
+
+        # Save dataframe to file and send it to the remote server if necessary
+        # Generate data file and send it to the remote (if necessary)
+        temp_dir = tempfile.mkdtemp(prefix='omniduct_hiveserver2')
+        tmp_fname = os.path.join(temp_dir, 'data_{}.csv'.format(time.time()))
+        logger.info('Saving dataframe to file... {}'.format(tmp_fname))
+        df.fillna(r'\N').to_csv(tmp_fname, index=False, header=False,
```

Comment:
does r'\N' get parsed as NULL by hive later?

Reply:
Yes. r'\N' is the character sequence interpreted as NULL (by default).